File: lasso_full.py
import time
import numpy as np

import numpy as np

def soft_thresholding(z, gamma):
    return np.sign(z) * np.maximum(np.abs(z) - gamma, 0.0)

def lasso_coordinate_descent(X, y, lambda_=0.1, max_iter=1000, tol=1e-4,lr=1.0):
    n_samples, n_features = X.shape
    beta = np.zeros(n_features)
    r = y - X @ beta  
    #X_norm_sq = np.sum(X ** 2, axis=0)/n_samples

    for iteration in range(max_iter):
        beta_old = beta.copy()

        for j in range(n_features):
            gk = -np.dot(X[:, j], r)/n_samples
            beta[j] = beta_old[j]-lr * gk
            beta[j] = soft_thresholding(beta[j], lambda_)
            delta = beta[j] - beta_old[j]
            r -= delta * X[:, j]

        if np.linalg.norm(beta - beta_old,ord=2) < tol:
            print(f"Converged in {iteration + 1} iterations.")
            break

    return beta





# Simulated data
np.random.seed(0)
n, p = 100, 1000
X = np.random.randn(n, p)
true_beta = np.zeros(p)
true_beta[:5] = [2, -1, 0.5, 0, 3]  # sparse true coefficients
y = X @ true_beta + np.random.randn(n) * 0.5

# Standardize X and center y
X -= X.mean(axis=0)
X /= X.std(axis=0)
y -= y.mean()

# Fit LASSO
start = time.time()
beta_est = lasso_coordinate_descent(X, y)
end = time.time()

print("Estimated coefficients:\n", np.round(beta_est, 3))
print("Time: ", (end - start) * 1000, "ms")

File: lasso_full.cpp
#include <iostream>
#include <Eigen/Dense>
#include <random>
#include <chrono>

using namespace Eigen;

double soft_thresholding(double z, double gamma) {
    return std::copysign(1.0, z) * std::max(std::abs(z) - gamma, 0.0);
}

VectorXd lasso_coordinate_descent(const MatrixXd& X, const VectorXd& y, double lambda=0.1, int max_iter=1000, double tol=1e-4, double lr=1.0) {
    int n_samples = X.rows();
    int n_features = X.cols();
    //VectorXd beta = VectorXd::Random(n_features);
    VectorXd beta = VectorXd::Zero(n_features);
    VectorXd residual = y - X * beta;

    for (int iteration = 0; iteration < max_iter; ++iteration) {
        VectorXd beta_old = beta;
        double max_delta = 0.0;

        for (int j = 0; j < n_features; ++j) {
            double gradient = -X.col(j).dot(residual) / n_samples;
            beta(j) = beta_old(j) - lr * gradient;
            beta(j) = soft_thresholding(beta(j), lambda);
            double delta = beta(j) - beta_old(j);
            residual -= X.col(j) * delta;
            if (std::abs(delta) > max_delta) {
                max_delta = std::abs(delta);
            }
        }

        if (max_delta < tol) {
            std::cout << "Converged in " << iteration + 1 << " iterations." << std::endl;
            return beta;
        }
    }
    std::cout << "Reached maximum iterations: " << max_iter << std::endl;
    return beta;
}


int main() {
    std::srand(0);
    int n = 100, p = 1000;
    MatrixXd X(n, p);
    VectorXd y(n);
    VectorXd true_beta = VectorXd::Zero(p);
    true_beta(0) = 2.0;
    true_beta(1) = -1.0;
    true_beta(2) = 0.5;
    true_beta(4) = 3.0;
    
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < p; ++j) {
            X(i, j) = std::rand() / double(RAND_MAX);
        }
        y(i) = X.row(i).dot(true_beta) + (std::rand() / double(RAND_MAX)) * 0.5;
    }
    
    VectorXd X_mean = X.colwise().mean();
    X = X.rowwise() - X_mean.transpose();
    VectorXd X_std = ((X.array().square().colwise().sum() / (n - 1)).sqrt());
    X = X.array().rowwise() / X_std.transpose().array();
    y = y.array() - y.mean();
    
    auto start = std::chrono::high_resolution_clock::now();
    VectorXd beta_est = lasso_coordinate_descent(X, y);
    auto end = std::chrono::high_resolution_clock::now();
    
    std::cout << "Estimated coefficients:\n" << beta_est.transpose() << std::endl;
    std::cout << "\nTime: " 
              << std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count()
              << " ms" << std::endl;
    
    return 0;
}